pkg/distributor/distributor.go

```diff
@@ -633,7 +638,46 @@ func (d *Distributor) HealthyInstancesCount() int {
 	return int(d.healthyInstancesCount.Load())
 }
 
-func extractSampleSeries(req *distributormodel.PushRequest) []*distributormodel.ProfileSeries {
+type groupsWithFingerprints struct {
+	labelSets    []phlaremodel.Labels
+	groups       []pprof.SampleGroup
+	fingerprints map[uint64][]int
+}
+
+func newGroupsWithFingerprints(n int) *groupsWithFingerprints {
+	return &groupsWithFingerprints{
+		labelSets:    make([]phlaremodel.Labels, 0, n),
+		groups:       make([]pprof.SampleGroup, 0, n),
+		fingerprints: make(map[uint64][]int),
+	}
+}
+
+func (g *groupsWithFingerprints) add(lbls phlaremodel.Labels, group pprof.SampleGroup) {
+	fp := lbls.Hash()
+	idxs, ok := g.fingerprints[fp]
+	if ok {
+		// fingerprint matches, check if the labels arethe same
+		for _, idx := range idxs {
+			if phlaremodel.CompareLabelPairs(g.labelSets[idx], lbls) == 0 {
+				// append samples to the group
+				g.groups[idx].Samples = append(g.groups[idx].Samples, group.Samples...)
```

Comment:
I'm afraid this may result in samples with identical stack traces, which is not expected in further processing (including the ingestion process). I can't foresee all the consequences, and the issue might be very obscure. I think it's safer to deduplicate them.

```go
type lazyGroup struct {
  sampleGroup pprof.SampleGroup
  // The map is only initialized when the group is being modified.
  // Key is the string representation (unsafe) of the sample stack trace.
  sampleMap map[string]*pprof.Sample
  labels    phlaremodel.Labels
}

// For reference (defined in pkg/pprof)
type SampleGroup struct {
	Labels  []*profilev1.Label
	Samples []*profilev1.Sample
}
```

Then, `groupsWithFingerprints` could be simplified to:
```go
type groupByFingerprint map[uint64][]*lazyGroup
```
That would be easier to read and (likely) a tiny bit faster

Reply:
For `groupByFingerprint` i also added a slice to keep the order intact.

See edb981689